app/projectsmodel.cpp

```diff
@@ -14,192 +8,582 @@
+      else if ( role == ProjectRemoteError ) return QVariant( project->mergin->remoteError );
+      return QVariant();
+    }
+  }
+}
+
+QModelIndex ProjectsModel::index( int row, int col, const QModelIndex &parent ) const
+{
+  Q_UNUSED( col )
+  Q_UNUSED( parent )
+  return createIndex( row, 0, nullptr );
+}
+
+QHash<int, QByteArray> ProjectsModel::roleNames() const
+{
+  QHash<int, QByteArray> roles;
+  roles[Roles::ProjectName]         = QStringLiteral( "ProjectName" ).toLatin1();
+  roles[Roles::ProjectNamespace]    = QStringLiteral( "ProjectNamespace" ).toLatin1();
+  roles[Roles::ProjectFullName]     = QStringLiteral( "ProjectFullName" ).toLatin1();
+  roles[Roles::ProjectId]           = QStringLiteral( "ProjectId" ).toLatin1();
+  roles[Roles::ProjectDirectory]    = QStringLiteral( "ProjectDirectory" ).toLatin1();
+  roles[Roles::ProjectIsLocal]      = QStringLiteral( "ProjectIsLocal" ).toLatin1();
+  roles[Roles::ProjectIsMergin]     = QStringLiteral( "ProjectIsMergin" ).toLatin1();
+  roles[Roles::ProjectSyncStatus]   = QStringLiteral( "ProjectSyncStatus" ).toLatin1();
+  roles[Roles::ProjectIsValid]      = QStringLiteral( "ProjectIsValid" ).toLatin1();
+  roles[Roles::ProjectFilePath]     = QStringLiteral( "ProjectFilePath" ).toLatin1();
+  roles[Roles::ProjectDescription]  = QStringLiteral( "ProjectDescription" ).toLatin1();
+  roles[Roles::ProjectPending]      = QStringLiteral( "ProjectPending" ).toLatin1();
+  roles[Roles::ProjectSyncProgress] = QStringLiteral( "ProjectSyncProgress" ).toLatin1();
+  roles[Roles::ProjectRemoteError]  = QStringLiteral( "ProjectRemoteError" ).toLatin1();
+  return roles;
+}
+
+int ProjectsModel::rowCount( const QModelIndex & ) const
+{
+  return mProjects.count();
+}
+
+void ProjectsModel::listProjects( const QString &searchExpression, int page )
+{
+  if ( mModelType == LocalProjectsModel )
+  {
+    listProjectsByName();
+    return;
+  }
+
+  mLastRequestId = mBackend->listProjects( searchExpression, modelTypeToFlag(), "", page );
 
-ProjectModel::ProjectModel( LocalProjectsManager &localProjects, QObject *parent )
-  : QAbstractListModel( parent )
-  , mLocalProjects( localProjects )
+  if ( !mLastRequestId.isEmpty() )
+    emit setModelIsLoading( true );
+}
+
+void ProjectsModel::listProjectsByName()
 {
-  findProjectFiles();
+  if ( mModelType != LocalProjectsModel )
+  {
+    return;
+  }
+
+  mLastRequestId = mBackend->listProjectsByName( projectNames() );
 
-  QObject::connect( &mLocalProjects, &LocalProjectsManager::localProjectAdded, this, &ProjectModel::addLocalProject );
+  if ( !mLastRequestId.isEmpty() )
+    emit setModelIsLoading( true );
 }
 
-ProjectModel::~ProjectModel() {}
+bool ProjectsModel::hasMoreProjects() const
+{
+  if ( mProjects.size() < mServerProjectsCount )
+    return true;
+
+  return false;
+}
+
+void ProjectsModel::fetchAnotherPage( const QString &searchExpression )
+{
+  listProjects( searchExpression, mPaginatedPage + 1 );
```

Comment:
mPaginatedPage is set to incremented value when response comes back to `onListProjectsfinished` (Response bears with it also paginated page). This way we only opt for `page + 1` and in case request would fail, we would still have the correct information about last paginated page